### PR TITLE
Ports TGMC autofire

### PR DESCRIPTION
**Pull request by Zonespace27**
# About the pull request
Tears out CM's autofire system and replaces it with TGMC's.
Shouldn't affect anything on the front-end beyond an unreported bug, and maybe making lag a bit better.

https://github.com/tgstation/TerraGov-Marine-Corps/pull/6064
https://github.com/tgstation/TerraGov-Marine-Corps/pull/6164

### Fire delay
F: `([^/])fire_delay = ([A-Z_1-90. *+-]+)`
R: `$1set_fire_delay($2)`

### Burst delay
F: `([^/])burst_delay = ([A-Z_1-90. *+-]+)`
R: `$1set_burst_delay($2)`

### Burst amount
F: `([^/])burst_amount = ([A-Z_1-90. *+-]+)`
R: `$1set_burst_amount($2)`

# Explain why it's good for the game
Autofire becoming less terrible is one of the things necessary to be able to add autofire to more guns.

# Changelog
:cl: BraveMole, Zonespace
fix: Smartgunners can no longer shoot their smartgun while aghosted
balance: The Uzi and Mac-15 now have full-auto
/:cl:

### CODE REVIEW

File: code/modules/projectiles/gun.dm [hunk 1]
```diff
@@ -1087,14 +1129,14 @@ and you're good to go.
...
-			return
+			return NONE
 			//If there's more to the attachment, it will be processed farther down, through in_chamber and regular bullet act.
 	/*
 	This is where burst is established for the proceeding section. Which just means the proc loops around that many times.
 	If burst = 1, you must null it if you ever RETURN during the for() cycle. If for whatever reason burst is left on while
 	the gun is not firing, it will break a lot of stuff. BREAK is fine, as it will null it.
 	*/
-	else if((flags_gun_features & GUN_BURST_ON) && burst_amount > 1)
+	else if((gun_firemode == GUN_FIREMODE_BURSTFIRE) && burst_amount > 1) //Zonenote: move to component?
```
Comment (mullenpaul) [inline code/modules/projectiles/gun.dm:1139]: I think the fire mode could be moved into a component applied to the gun

File: code/modules/projectiles/gun.dm [hunk 2]
```diff
@@ -1635,7 +1677,7 @@ not all weapons use normal magazines etc. load_into_chamber() itself is designed
 
 	// Full auto fucks your scatter up big time
 	// Note that full auto uses burst scatter multipliers
-	if(flags_gun_features & GUN_FULL_AUTO_ON)
+	if(gun_firemode == GUN_FIREMODE_AUTOMATIC)
```
Comment (mullenpaul) [inline code/modules/projectiles/gun.dm:1680]: This makes me think that capturing the firemodes in a datum would be for the best. We can then would have much more flexibility in how things worked

**Comment by mullenpaul** [comment]
Whilst it would be neat to refactor firemodes into datums - I think it could complicate the code at this point. I would generally think that we should port the code over -as is- and then make improvements on it

File: code/modules/projectiles/gun.dm (hunk 1 above)
Comment (Zonespace27) [reply code/modules/projectiles/gun.dm:1139]: What's gained from that?

File: code/__DEFINES/autofire.dm [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// Controls how many buckets should be kept, each representing a tick. Max is ten seconds, to have better perf.
+#define AUTOFIRE_BUCKET_LEN (world.fps * 10)
+/// Helper for getting the correct bucket
+#define AUTOFIRE_BUCKET_POS(next_fire) (((round((next_fire - SSautomatedfire.head_offset) / world.tick_lag) + 1) % AUTOFIRE_BUCKET_LEN) || AUTOFIRE_BUCKET_LEN)
```
Comment (fira) [inline code/__DEFINES/autofire.dm:4]: think these should be in the SS file, same way as SStimer does it, having private bits in separate define files will just confuse people coming across it

File: code/modules/projectiles/guns/flamer/flamer.dm [hunk 4]
```diff
@@ -360,6 +362,12 @@
 		return TRUE
 	return FALSE
 
+/obj/item/weapon/gun/flamer/M240T/auto // With NEW advances in science, we've learned how to drain a pyro's tank in 6 seconds, or your money back!
+	name = "\improper M240-T2 incinerator unit"
+	desc = "A prototyped model of the M240-T incinerator unit, it was discontinued after its automatic mode was deemed too expensive to deploy in the field."
+	start_semiauto = FALSE
+	start_automatic = TRUE
```
[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/modules/projectiles/guns/flamer/flamer.dm:369]: Admin spawn modification of the spec flamer that takes advantage of the fact that i made full auto work for flamers

File: code/modules/projectiles/gun.dm [hunk 5]
```diff
@@ -724,14 +762,11 @@
...
 
-	if(flags_gun_features & GUN_FULL_AUTO_ON)
-		ADD_TRAIT(user, TRAIT_OVERRIDE_CLICKDRAG, TRAIT_SOURCE_WEAPON)
-
 	return 1
 
 /obj/item/weapon/gun/unwield(mob/user)
 	. = ..()
-	REMOVE_TRAIT(user, TRAIT_OVERRIDE_CLICKDRAG, TRAIT_SOURCE_WEAPON)
+	//REMOVE_TRAIT(user, TRAIT_OVERRIDE_CLICKDRAG, TRAIT_SOURCE_WEAPON)
```
Comment (harryob) [inline code/modules/projectiles/gun.dm:769]: remove?

File: code/modules/projectiles/gun_helpers.dm [hunk 6]
```diff
@@ -681,80 +658,80 @@ As sniper rifles have both and weapon mods can change them as well. ..() deals w
...
-				COMSIG_CLIENT_LMB_DOWN,
-				COMSIG_CLIENT_LMB_UP,
-				COMSIG_CLIENT_LMB_DRAG,
-			))
-
-			to_chat(user, SPAN_NOTICE("[icon2html(src, user)] You set [src] to single fire mode."))
-			return
+/obj/item/weapon/gun/proc/remove_firemode(removed_firemode, mob/user)
+	switch(length(gun_firemode_list))
+		if(0, 1)
```
Comment (harryob) [inline code/modules/projectiles/gun_helpers.dm:698]: looks a bit odd with only one condition?

File: code/modules/projectiles/gun.dm (hunk 5 above)
Comment (Zonespace27) [reply code/modules/projectiles/gun.dm:769]: removed

File: code/modules/projectiles/gun_helpers.dm (hunk 6 above)
Comment (Zonespace27) [reply code/modules/projectiles/gun_helpers.dm:698]: changed

File: code/modules/projectiles/gun_helpers.dm [hunk 7]
```diff
@@ -681,80 +658,80 @@ As sniper rifles have both and weapon mods can change them as well. ..() deals w
...
+	if(start_semiauto)
+		gun_firemode_list |= GUN_FIREMODE_SEMIAUTO
+
+	if(burst_amount > BURST_AMOUNT_TIER_1)
+		gun_firemode_list |= GUN_FIREMODE_BURSTFIRE
+
+	if(start_automatic)
+		gun_firemode_list |= GUN_FIREMODE_AUTOMATIC
+
+	switch(length(gun_firemode_list))
```
[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/modules/projectiles/gun_helpers.dm:717]: changed

File: code/datums/components/autofire/autofire.dm [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	///Callback to ask the parent to reset its firing vars
+	var/datum/callback/callback_reset_fire
+	///Callback to ask the parent to fire
+	var/datum/callback/callback_fire
+	///Callback to ask the parent to display ammo
+	var/datum/callback/callback_display_ammo
+	///Callback to set parent's fa_firing
+	var/datum/callback/callback_set_firing
+
+/datum/component/automatedfire/autofire/Initialize(_auto_fire_shot_delay = 0.3 SECONDS, _burstfire_shot_delay, _burst_shots_to_fire = 3, _fire_mode = GUN_FIREMODE_SEMIAUTO, datum/callback/_callback_bursting, datum/callback/_callback_reset_fire, datum/callback/_callback_fire, datum/callback/_callback_display_ammo, datum/callback/_callback_set_firing)
```
Comment (harryob) [inline code/datums/components/autofire/autofire.dm:29]: prefer leaving the names without _s and just using src.auto_fire_shot_delay, per https://github.com/cmss13-devs/cmss13/blob/master/.github/guides/STYLES.md#when-passing-vars-through-new-or-initializes-arguments-use-srcvar

File: code/datums/components/autofire/autofire.dm [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+/datum/component/automatedfire/autofire/Destroy(force, silent)
+	QDEL_NULL(callback_fire)
+	QDEL_NULL(callback_reset_fire)
+	QDEL_NULL(callback_bursting)
+	QDEL_NULL(callback_display_ammo)
+	QDEL_NULL(callback_set_firing)
+	return ..()
+
+///Setter for fire mode
+/datum/component/automatedfire/autofire/proc/modify_fire_mode(datum/source, _fire_mode)
```
Comment (harryob) [inline code/datums/components/autofire/autofire.dm:58]: i suppose the same as above applies to these

**Review by harryob — CHANGES_REQUESTED** [review]
one last nit

**Review by harryob — APPROVED** [review]

**Pull request merged**